### PR TITLE
feat: support for multiple TES versions

### LINKED ISSUE #48: Add support for multiple TES versions in the test reports
Currently the tes-compliance-suite report only supports TES v1.0. The ability to support multiple test reports would be beneficial given the release of TES v1.1, along with future versions.

Proposed behavior:
- One page/report for each TES version
  - TES v1.0: `report-tes-1.0.0.html`
  - TES v1.1: `report-tes-1.1.0.html`
- Landing page would be the earliest TES version passed to the compliance suite 
  - `index.html` → `report-tes-1.0.0.html`

---

Happy to take this on while following your recommendations for this @LakiG! Will push a PR for review for this today and update here if that works.

### PR DESCRIPTION
**Pull request by lbeckman314**
# Overview

Resolves #48 (Add support for multiple TES versions in the test reports)

Happy to update or remove anything based on your discretion!

### Multiple `--version` Flags

Allow for multiple `--version` flags to be passed to `cli.py`. Example of testing a TES server running on "http://localhost:8000/":

```sh
tes-compliance-suite report --version "1.0.0" --version "1.1.0" --server "http://localhost:8000/" --tag "get service info test" --serve
```

### Multiple Report Pages

Add a report page for every TES version passed in as above.

**TES v1.0 Report**

tes-v1 0

**TES v1.1 Report**

tes-v1 1

### CODE REVIEW

**Comment by lbeckman314** [comment]
Ah, working on the failing tests now!

File: compliance_suite/cli.py [hunk 1]
```diff
@@ -20,14 +20,14 @@ def main() -> None:
 
 @main.command(help='Run TES compliance tests against the servers')
 @click.option('--server', '-s', help='server URL on which the compliance tests are run. Format - https://<url>/')
-@click.option('--version', '-v', help='TES version. Example - "1.0.0"')
+@click.option('--version', '-v', 'versions', multiple=True, help='TES version. Example - "1.0.0"')
```
Comment (uniqueg) [inline compliance_suite/cli.py]: Maybe change the description, too? E.g., "One or more TES versions. Example [...]"

File: compliance_suite/cli.py [hunk 2]
```diff
@@ -38,41 +38,55 @@ def report(server: str,
 
     Args:
         server (str): The server URL on which the compliance suite will be run. Format - https://<url>/
-        version (str): The compliance suite will be run against this TES version. Example - "1.0.0"
+        version (List[str]): The compliance suite will be run against this TES version. Example - "1.0.0"
         tag (List[str]): The list of the tags for which the compliance suite will be run. Default - All
         output_path (str): The output path to store the JSON compliance report
         serve (bool): If true, runs a local server and displays the JSON report in webview
-        port (int): Set the local server port. Default - 16800
+        port (int): Set the local server port. Default - 15800
```
Comment (uniqueg) [inline compliance_suite/cli.py]: Any particular reason for the port change?

File: compliance_suite/cli.py [hunk 3]
```diff
@@ -38,41 +38,55 @@ def report(server: str,
 
     Args:
         server (str): The server URL on which the compliance suite will be run. Format - https://<url>/
-        version (str): The compliance suite will be run against this TES version. Example - "1.0.0"
+        version (List[str]): The compliance suite will be run against this TES version. Example - "1.0.0"
```
Comment (uniqueg) [inline compliance_suite/cli.py]: Again, would be good to change the description here as well so that it is clear that it can be a list of versions.

File: compliance_suite/cli.py [hunk 4]
```diff
@@ -38,41 +38,55 @@ def report(server: str,
...
+        job_runner = JobRunner(server, version, tag)
+        job_runner.run_jobs()
+
+        json_report = job_runner.generate_report()
+
+        # Store the report in given output path
+        if output_path is not None:
+            logger.info(f"Writing JSON Report on directory {output_path}")
+            with open(os.path.join(output_path, f"report-tes-{version}.json"), "w+") as output:
+                output.write(json_report)
```
Comment (uniqueg) [inline compliance_suite/cli.py]: It seems to me that in this way, _all_ available tests for a given version run for every version the user supplies. However, in many cases, tests will be applicable for multiple versions. And in these cases, the test results could be reported for multiple versions, too, to minimize runtime and save resources.

File: compliance_suite/report_server.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 import http.server
 import json
 import os
+from typing import List
```
Comment (uniqueg) [inline compliance_suite/report_server.py]: Please follow import ordering used in library. Not sure how @LakiG deals with `from x import y` versus `import x`, but it seems to me that at least within each of these, there is alphabetical sorting according to the package name.

File: compliance_suite/report_server.py [hunk 6]
```diff
@@ -64,7 +67,7 @@ def start_local_server(
         http_handler = http.server.SimpleHTTPRequestHandler
         self.local_server = socketserver.TCPServer(("", port), http_handler)
         logger.info(f"Starting a local server at  http://localhost:{port}")
-        webbrowser.open(f"http://localhost:{port}")
+        webbrowser.open(f"http://localhost:{port}/report-tes-{self.versions[0]}.html")
```
Comment (uniqueg) [inline compliance_suite/report_server.py]: Is this how the default version to show is set? If so, I'm not sure the behavior is as you say, i.e., that the version the user specifies _first_ is the one that is shown. Rather, it would, I think, be always the smallest version, because of `versions.sort()` in the CLI module.

However, personally I'm actually not quite sure there's a lot of benefit to allow the user specifying the order of the reports (and if so, that behavior should be documented somewhere). Rather, I would always just start with the latest version.

File: compliance_suite/web/index.html
Comment (uniqueg) [inline compliance_suite/web/index.html:1]: Was this deliberately deleted? Is it maybe referred to in the docs, as an example? If it was deleted because it was inadvertently added, perhaps it should also be added to `.gitignore`?

File: compliance_suite/web/partials/navbar.html [hunk 7]
```diff
@@ -1,3 +1,6 @@
 <nav id="navbar" class="navbar navbar-expand-lg navbar-light bg-light">
-    <a id="navbar-title" class="navbar-brand" href="#">GA4GH TES v0.1.0 Compliance Report</a>
+    <a id="navbar-title" class="navbar-brand" href="#">GA4GH TES Compliance Report</a>
+    {% for version in versions %}
+        <a id="navbar-title" class="navbar-brand" href="report-tes-{{version}}.html">TES v{{version}}</a>
+    {% endfor %}
```
Comment (uniqueg) [inline compliance_suite/web/partials/navbar.html]: I think this is good opportunity to make the test runner more generic by not only not hardcoding the version, but also not hardcoding the API _name_ (here: "GA4GH TES"). Also, "Compliance" is perhaps too specific, I would rather go with "Test Report".

File: compliance_suite/web/views/report.html [hunk 8]
```diff
@@ -1,9 +1,8 @@
-{% include "partials/header.html" %}
-{% include "partials/navbar.html" %}
-
+{% extends "views/base.html" %}
+{% block report %}
 <div class="container">
 
-  <h2>TES Compliance Report</h2>
+  <h2>TES v{{version}} Compliance Report</h2>
```
Comment (uniqueg) [inline compliance_suite/web/views/report.html]: Same as above: If not too difficult, try to get rid of "TES" as well.

File: compliance_suite/web/web_report.json
Comment (uniqueg) [inline compliance_suite/web/web_report.json:1]: Same as above: just checking whether deletion was deliberate and whether it may make sense to add this file to `.gitignore`

File: tests/create_task_streamable.yml [hunk 9]
```diff
@@ -26,7 +26,7 @@ jobs:
             ]
           }
         ],
-        "input": [
+        "inputs": [
```
Comment (uniqueg) [inline tests/create_task_streamable.yml:29]: I guess this is technically a part of another issue.

File: tests/create_task_output.yml
Comment (uniqueg) [inline tests/create_task_output.yml:1]: Is this a new test? If so, maybe it should go to a different PR as it has nothing to do with the versions?

Also, is this really a test specific to TES 1.1.0? It seems to me that it should also be applicable to TES 1.0.0, but I may be wrong.

**Review by uniqueg — CHANGES_REQUESTED** [review]
Good stuff! A couple of comments, some just to clarify things I'm not sure I understand.

There is, I think, also another issue that may be worth thinking about: If (and only if!) we can assume that the test runner will only be used with semantically versioned APIs, then I think supplying multiple minor or patch versions is pointless, because it will be expected that _every_, say, `1.0.0` test will be expected to yield the same result for `1.0.0` and `1.x.y` implementations. The same, btw, goes for the version tags. Rather than tagging them with multiple minor versions, tagging them with multiple major versions should be enough.

And while I think that it is probably too strong an assumption that all tested APIs will be semantically versioned, this assumption may already be hardcoded in the current test runner. So I see three options here:
1. Clearly state that assumption and update the test runner accordingly, for the version labeling of tests, i.e., only supply the minimum version for each _major_ version release of the API, e.g., `1.1.1` and `2.0.1`, for the selection of which tests to run (all tests with the same major and equal or lower minor/patch version combination), and for this PR, i.e., only keep the highest minor version for each major versions and log a warning if the user supplies any redundant minor version, e.g., if they supply `1.0.1`, `1.1.1`, `2.0.0` and `2.0.1`, then tests would only be run for `1.1.1` and `2.0.1`.
2. Refactor/modify all code that assumes that an API versions follows SemVer.
3. Allow the user to indicate whether an API follows SemVer or not (default: `True`), and depending on that information, either make use of the implications of SemVer adoption to to minimize test runs, or not.

Personally I tend to favor the last two options (potentially supporting more APIs). Ideally the last one (best of both worlds), but it may add too much complexity. What do you both think?

File: compliance_suite/cli.py (hunk 2 above)
Comment (lbeckman314) [reply compliance_suite/cli.py]: Ah hopefully this reconciles the existing default port used by Click, happy to change it either way! 

```py
@click.option('--port', default=15800, help='port at which the compliance report is served')
```

https://github.com/elixir-cloud-aai/tes-compliance-suite/blob/dev/compliance_suite/cli.py#L27

Comment (uniqueg) [reply compliance_suite/cli.py]: No, it's fine with me, if @LakiG doesn't have a particular reason to keep `16800`.

Comment (LakiG) [reply compliance_suite/cli.py]: Thanks for pointing that out.

File: compliance_suite/report_server.py (hunk 5 above)
Comment (LakiG) [reply compliance_suite/report_server.py]: Yeah, there doesnt seem to be any guideline about this. Ref - https://peps.python.org/pep-0008/#imports
But within the same import group, it would be better to separate them. Example:

`import x` _All alphabetically sorted_
`from x import y` _All alphabetically sorted_

File: compliance_suite/web/index.html
Comment (LakiG) [reply compliance_suite/web/index.html:1]: Although its not being referred, it could serve as an example file.

We can refer to this in the docs as well when we have developer docs specifically.

File: compliance_suite/web/partials/navbar.html (hunk 7 above)
Comment (LakiG) [reply compliance_suite/web/partials/navbar.html]: I agree, @lbeckman314 added a new field `name`. That can be used here.

**Comment by uniqueg** [comment]
@LakiG: Did you review the PR?
@lbeckman314: Will you tag me/us for re-review when you are done with changes from your side?

File: compliance_suite/cli.py [hunk 10]
```diff
@@ -49,30 +52,48 @@ def report(server: str,
...
+    logger.info(f"Input version(s) - {versions}")
 
-    # Store the report in given output path
-    if output_path is not None:
-        logger.info(f"Writing JSON Report on directory {output_path}")
-        with open(os.path.join(output_path, "report.json"), "w+") as output:
-            output.write(json_report)
+    logger.info(f"Input name - {name}")
+
+    any_fail = False
```
Comment (LakiG) [inline compliance_suite/cli.py:66]: Please add the variable annotations for new variables.

File: compliance_suite/cli.py [hunk 11]
```diff
@@ -49,30 +52,48 @@ def report(server: str,
...
     if serve is True:
-        report_server = ReportServer(os.path.join(os.getcwd(), "compliance_suite", "web"))
+        report_server = ReportServer(os.path.join(os.getcwd(), "compliance_suite", "web"), versions, name)
         report_server.serve_thread(port, uptime)
 
+        exit()
+
+    # If report has failed tests, exit with error code
+    if any_fail is True:
+        exit(1)
```
Comment (LakiG) [inline compliance_suite/cli.py:95]: In a scenario where `serve` is `True` and there were failed cases, shouldn't we get `exit(1)`?

File: compliance_suite/job_runner.py [hunk 12]
```diff
@@ -55,6 +55,9 @@ def __init__(self, server: str, version: str, tags: List[str]):
         }
         self.report: Any = None
 
+    def get_test_status(self) -> Dict[str, List[str]]:
+        return self.test_status
```
Comment (LakiG) [inline compliance_suite/job_runner.py:59]: Maybe we can add the docstring for this new method

File: compliance_suite/report_server.py [hunk 13]
```diff
@@ -4,24 +4,30 @@
...
+from typing import (
+    Any,
+    List
+)
 import webbrowser
 
 import jinja2 as j2
 
 from compliance_suite.functions.log import logger
+from pathlib import Path
```
Comment (LakiG) [inline compliance_suite/report_server.py:24]: We can move this to the first import group as it is a standard in-built library.

File: compliance_suite/report_server.py [hunk 14]
```diff
@@ -31,22 +37,31 @@ def __init__(self, web_dir: str):
         self.cwd: str = os.getcwd()     # Store the current location
         self.local_server: Any = None
         self.web_dir: str = web_dir
+        self.versions: List[str] = versions
+        self.name = name
```
Comment (LakiG) [inline compliance_suite/report_server.py:41]: The variable notation is needed here as well.

File: unittests/test_report_server.py [hunk 15]
```diff
@@ -4,24 +4,25 @@
 """
 
 import os
-import unittest
 from unittest.mock import (
     MagicMock,
     patch
 )
-
```
Comment (LakiG) [inline unittests/test_report_server.py:12]: This line break is required to separate the import groups.

File: compliance_suite/cli.py (hunk 4 above)
Comment (LakiG) [reply compliance_suite/cli.py]: If a test is executed by any of the versions, we should not repeat it again as the result would be the same every time.
We'll need to explore how to store the state of executed tests.

Another thing to note, when generating the report, we'll need the copy and append the common test result of the first version to the other versions. Otherwise, they'll miss out on tests.

**Review by LakiG — CHANGES_REQUESTED** [review]
You can modify the PR name since its not only applicable to test reports. We are adding support for multiple versions overall.

**Comment by LakiG** [comment]
> Personally I tend to favor the last two options (potentially supporting more APIs). Ideally the last one (best of both worlds), but it may add too much complexity. What do you both think?

Currently, we are following the second point where we assume all API versions maintain SemVer. Though we earlier had a discussion on this, regarding the base URL to be used for non-SemVer APIs and couldn't reach a conclusion. We can add support for those if we face any such API later.

Regarding the point that a major version would give the same test results, it depends on how the API developers have maintained their versions. For example, if we see TES only, some new features were added in the minor version 1.1.0. The new tests would fail for the previous minor version 1.0.0, so essentially the same major version is giving different results.
 
cc - @uniqueg @lbeckman314

**Comment by lbeckman314** [comment]
Apologies for the slow updates -- we've been using this branch of tes-compliance-suite to update[ Funnel's v1.0 and v1.1 compliance status](https://ohsu-comp-bio.github.io/funnel-compliance/report-TES-1.1.0.html) in preparation for it's newest release.

I'll plan on rebasing onto the latest updates in [elixir-cloud-aai:dev](https://github.com/elixir-cloud-aai/tes-compliance-suite/tree/dev) today to make any potential merges or recommendations easier!

Also happy to close this PR if development is using a different strategy for managing different TES versions.

Just a side note but we were able to briefly show off your all's work on the tes-compliance-suite this morning on the [GA4GH Cloud Work Stream](https://docs.google.com/document/d/1hayvWLIoymomPI9oXcaTZirn5YxFv1cYAs70zyvlvnA/edit#heading=h.o1ijhfdu9zv) meeting!

**Comment by uniqueg** [comment]
Hi @lbeckman314 - thanks for the update! We'd be happy to merge this PR upstream once you have updated/rebased it, so please let us know when that is :)

Also, thanks for demonstrating the compliance suite at the GA4GH meeting :rocket:

**Comment by vsmalladi** [comment]
@lbeckman314 @LakiG  can we merge this in?

**Comment by lbeckman314** [comment]
Ah, looking over this PR it might be best to close it and open separate ones for the following features:
- Report publishing for multiple versions
- Exit code 1 on test failure (for Github Action status reporting)

I will go ahead and close this PR and create these new more specific ones for review!
